src/qspi.rs

```diff
@@ -0,0 +1,783 @@
+        }
+
+        // Write data to the FIFO
+        if let Some((data, _)) = command.data {
+            for byte in data {
+                while self.qspi.sr.read().ftf().bit_is_clear() {}
+                unsafe {
+                    core::ptr::write_volatile(self.qspi.dr.as_ptr() as *mut u8, *byte);
+                }
+            }
+        }
+
+        while self.qspi.sr.read().tcf().bit_is_clear() {}
+
+        self.qspi.fcr.write(|w| w.ctcf().set_bit());
+
+        if command.double_data_rate {
+            self.qspi.cr.modify(|_, w| {
+                w.sshift()
+                    .bit(self.config.sample_shift == SampleShift::HalfACycle)
+            });
+        }
+        Ok(())
+    }
+
+    /// Put the QSPI peripheral into memory mapped mode. Returns a slice to the
+    /// memory mapped region.
+    /// Provide the command for the read operation for your flash chip.
+    pub fn memory_mapped<'a>(
+        &'a mut self,
+        command: QspiMemoryMappedConfig,
+    ) -> Result<MemoryMapped<'a, PINS>, QspiError> {
+        if self.is_busy() {
+            return Err(QspiError::Busy);
+        }
+
+        // If double data rate change shift
+        if command.double_data_rate {
+            self.qspi.cr.modify(|_, w| w.sshift().bit(false));
+        }
+        while self.is_busy() {}
+
+        // Clear the transfer complete flag.
+        self.qspi.fcr.modify(|_, w| w.ctcf().set_bit());
+
+        let mut abmode: u8 = 0;
+        let mut absize: u8 = 0;
+        let mut imode: u8 = 0;
+        let mut instruction: u8 = 0;
+
+        if let Some((inst, mode)) = command.instruction {
+            imode = mode as u8;
+            instruction = inst;
+        }
+
+        // Write alternate bytes
+        if let Some((a_bytes, mode)) = command.alternate_bytes {
+            abmode = mode as u8;
+            absize = a_bytes.len() as u8 - 1;
+
+            self.qspi.abr.write(|w| {
+                let mut reg_byte: u32 = 0;
+                for (i, element) in a_bytes.iter().rev().enumerate() {
+                    reg_byte |= (*element as u32) << (i * 8);
+                }
+                unsafe { w.alternate().bits(reg_byte) }
+            });
+        }
+
+        self.qspi.ccr.modify(|_, w| unsafe {
+            w.fmode()
+                .bits(0b11 /* Memory mapped mode */)
+                .admode()
+                .bits(command.address_mode as u8)
+                .adsize()
+                .bits(self.config.address_size as u8)
+                .abmode()
+                .bits(abmode)
+                .absize()
+                .bits(absize)
+                .ddrm()
+                .bit(command.double_data_rate)
+                .dcyc()
+                .bits(command.dummy_cycles)
+                .dmode()
+                .bits(command.data_mode as u8)
+                .imode()
+                .bits(imode)
+                .instruction()
+                .bits(instruction)
+        });
```

Comment:
You can do this more compact
```suggestion
        self.qspi.ccr.modify(|_, w| unsafe {
            w.fmode().bits(0b11 /* Memory mapped mode */);
            w.admode().bits(command.address_mode as u8);
            w.adsize().bits(self.config.address_size as u8);
            w.abmode().bits(abmode);
            w.absize().bits(absize);
            w.ddrm().bit(command.double_data_rate);
            w.dcyc().bits(command.dummy_cycles);
            w.dmode().bits(command.data_mode as u8);
            w.imode().bits(imode);
            w.instruction().bits(instruction)
        });```